go/interpreter/lfvm/instruction_logger.go

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package lfvm
+
+import (
+	"fmt"
+	"io"
+)
+
+// loggingRunner is a runner that logs the execution of the contract code to
+// an io.Writer. If no writter is provided with NewLoggingRunner, the log will
+// be written to os.Stderr.
+type loggingRunner struct {
+	log io.Writer
+}
+
+// newLogger creates a new logging runner that writes to the provided
+// io.Writer.
+func newLogger(writer io.Writer) loggingRunner {
+	return loggingRunner{log: writer}
+}
+
+func (l loggingRunner) run(c *context) (status, error) {
+	status := statusRunning
+	var err error
+	for status == statusRunning {
+		// log format: <op>, <gas>, <top-of-stack>\n
+		if int(c.pc) < len(c.code) {
+			top := "-empty-"
+			if c.stack.len() > 0 {
+				top = c.stack.peek().ToBig().String()
+			}
+			if l.log != nil {
+				_, err = l.log.Write([]byte(fmt.Sprintf("%v, %d, %v\n", c.code[c.pc].opcode, c.gas, top)))
+				if err != nil {
+					return status, err
```

Comment:
I am conflicted here, and I think we may have a problem because of the error refactoring. 
there is no statusError anymore, and we assume status is an error if there is an error. This is logically correct because all the errors propagated by the vm are runtime errors product of the input. 

The error in this line is not, i.e. this can happen when the filesystem is full. The code calling this error treats the error as an execution violation, sets the result status to Reverted and consumes all the gas. The user will never see that the execution stopped and the error will be lost. 
How can we solve this?

Reply:
@LuisPH3: I agree, those two categories of errors should be separated.

I see two options:
- return three result values: the status, an "error" status which is an enum but not really an error, and an error;
- return two results: the status which is extended with more error states, and an error that is always nil in the vanilla runner but used for this situation here

I think I would prefer the second one, reserving "errors" for real errors, not valid execution results.

Reply:
yes, this may require a little work, but it would be more solid. Let me coordinate this work with Facu

Reply:
if you agree, I would leave it as is in this PR, and handle `config.run` return in a different PR.

Reply:
:+1:  I will start the fix shortly today. 
It is not really in the scope of the logger, therefore I am fine with this for now.

Reply:
TBD: #758